Compiler Bug With Uninitialized Generic Struct Fields

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Gentoo Base System release 2.8"
Processor: 128 cpus, 64bit, little endian, AMD Ryzen Threadripper 3990X 64-Core Processor
CC version: cc (Gentoo 11.2.0 p1) 11.2.0

getwd: /home/valczir/dev/vipar
vmodules: /home/valczir/.vmodules
vroot: /home/valczir/.local/share/v
vexe: /home/valczir/.local/share/v/v
vexe mtime: 2021-11-06 16:05:16
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 8be64ef

Git version: git version 2.33.1
Git vroot status: weekly.2021.44-24-g8be64ef8
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg default_slice_field.v`
```v
struct Foo<T> {
	foo T
}

struct Bar<T> {
mut:
	foos []Foo<T>
}

fn (mut b Bar<T>) add(v T) {
	b.foos << Foo<T>{
		foo: v
	}
}

fn main() {
	mut bar := Bar<string>{}
	bar.add('bar')
	println(bar)
}

```

**What did you expect to see?**

~A compile-time error telling me that `Bar<string>` cannot be constructed without a value for the `foos` field.  This seems like the most consistent way to deal with the issue, since variables in v can't be initialized without a value - it would make sense to me if the same restriction applied to struct fields.~

^^^ Never mind that, I found the previous discussions about default values on struct fields.

Either the default value for  a slice should be appendable (as in `go`), or I should get a stack trace showing that `foos` was not initialized and cannot be appended to.


**What did you see instead?**
```
==================
/tmp/valczir/v/default_slice_field.1573690348668029940.tmp.c:828: error: ';' expected (got "<")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/ast/table.v
+++ b/vlib/v/ast/table.v
@@ -1661,16 +1661,14 @@ pub fn (mut t Table) generic_insts_to_concrete() {
 						generic_names := parent_info.generic_types.map(t.get_type_symbol(it).name)
 						for i in 0 .. fields.len {
 							if fields[i].typ.has_flag(.generic) {
-								sym := t.get_type_symbol(fields[i].typ)
-								if sym.kind == .struct_ && fields[i].typ.idx() != info.parent_idx {
+								if fields[i].typ.idx() != info.parent_idx {
 									fields[i].typ = t.unwrap_generic_type(fields[i].typ,
 										generic_names, info.concrete_types)
-								} else {
-									if t_typ := t.resolve_generic_to_concrete(fields[i].typ,
-										generic_names, info.concrete_types)
-									{
-										fields[i].typ = t_typ
-									}
+								}
+								if t_typ := t.resolve_generic_to_concrete(fields[i].typ,
+									generic_names, info.concrete_types)
+								{
+									fields[i].typ = t_typ
 								}
 							}
 						}

--- a/vlib/v/tests/generics_nested_struct_init_test.v
+++ b/vlib/v/tests/generics_nested_struct_init_test.v
@@ -1,0 +1,25 @@
+struct Foo<T> {
+	foo T
+}
+
+struct Bar<T> {
+mut:
+	foos []Foo<T>
+}
+
+fn (mut b Bar<T>) add(v T) {
+	b.foos << Foo<T>{
+		foo: v
+	}
+}
+
+fn test_nested_generics_struct_init() {
+	mut bar := Bar<string>{}
+	bar.add('bar')
+	println(bar)
+
+	result := '$bar'
+	assert result.contains('Bar<string>{')
+	assert result.contains('foos: [Foo<string>{')
+	assert result.contains("foo: 'bar'")
+}

